### PR TITLE
DOCS-2653: Update the training script docs

### PR DESCRIPTION
**Pull request by npentrel**
@tahiyasalam can you review the template please to make sure I didn't mess that up? 

There are more changes I'll be making but I think that's the biggest one.

Also: Is it only setup.py that’s needed in addition? Because there’s also:

- PKG-INFO
- model
    - `__init__.py`
    - setup.cfg
    - setup.py
- training.egg-info
    - PKG-INFO
    - SOURCES.txt
    - dependency_links.txt
    - requires.txt
    - top_level.txt

### CODE REVIEW

File: docs/services/ml/upload-training-script.md [hunk 1]
```diff
@@ -10,48 +10,231 @@ no_service: true
...
+   )
+   ```
+
+   Ensure you add additional required packages on line 11.
+
+1. Inside the top level folder (in this example <file>my-training</file>), create a folder called <file>model</file> and create an empty file called <file>\_\_init\_\_.py</file> inside it.
+1. Inside the <file>model</file> folder, create a file called <file>training.py</file>.
+
+   {{< alert title="Using Viam APIs in a training script" color="note" >}}
+   If you need to access any of the [Viam APIs](/appendix/apis/) within a custom training script, you can use the environment variables `API_KEY` and `API_KEY_ID` to establish a connection.
```
Comment (tahiyasalam) [inline docs/services/ml/upload-training-script.md:48]: Do you think it's worth having a snippet for the connection?

```
async def connect() -> ViamClient:
    """Returns a authenticated connection to the ViamClient for the requested org associated with the submitted training job."""
    # The API key and key ID can be accessed programmatically, using the environment variable API_KEY and API_KEY_ID.
    # The user does not need to supply the API keys, they are provided automatically when the training job is submitted.
    dial_options = DialOptions.with_api_key(
        os.environ.get("API_KEY"), os.environ.get("API_KEY_ID")
    )
    return await ViamClient.create_from_dial_options(dial_options)
```

File: docs/services/ml/upload-training-script.md [hunk 2]
```diff
@@ -10,48 +10,231 @@ no_service: true
...
+   # bounding box annotations
+   # Used for training object detection models
+   def parse_filenames_and_bboxes_from_json(
+       filename: str,
+       all_labels: ty.List[str],
+   ) -> ty.Tuple[ty.List[str], ty.List[str], ty.List[ty.List[float]]]:
+       """Load and parse JSON file to return image filenames
+       and corresponding labels with bboxes.
+       Args:
+           filename: JSONLines file containing filenames and bboxes
```
Comment (tahiyasalam) [inline docs/services/ml/upload-training-script.md:125]: This is my bad but args should all_labels: list of all N_LABELS here as well :)

File: docs/services/ml/upload-training-script.md [hunk 3]
```diff
@@ -10,48 +10,231 @@ no_service: true
...
+       labels: ty.List[str], model_type: str, input_shape: ty.Tuple[int, int, int]
+   ) -> Model:
+       """Builds and compiles a model
+       Args:
+           labels: list of string lists, where each string list contains up to N_LABEL labels associated with an image
+           model_type: string single_label or multi_label
+           input_shape: 3D shape of input
+       """
+
+       # TODO: Add logic to build and compile model
```
Comment (tahiyasalam) [inline docs/services/ml/upload-training-script.md:167]: Is this supposed to be a TODO? If so, I do think it would be helpful to have a working example, but I will defer to you

File: docs/services/ml/upload-training-script.md [hunk 4]
```diff
@@ -124,154 +307,45 @@ Dataset JSON files for image datasets with bounding box labels and classificatio
...
-    model = tf.keras.Model(x, y)
-
-    model.compile(
-        loss=loss_fnc,
-        optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
-        metrics=[metrics],
-    )
-    return model
-```
+As an example, you can refer to the logic from <file>model/training.py</file> from this [example classification training script](https://app.viam.com/packages/e76d1b3b-0468-4efd-bb7f-fb1d2b352fcb/custom-training-classification/ml_training/latest/e76d1b3b-0468-4efd-bb7f-fb1d2b352fcb) that trains a classification model using TensorFlow and Keras.
```
Comment (tahiyasalam) [inline docs/services/ml/upload-training-script.md:318]: I was going to make a follow-up ticket after the epic was completed, but I think we can just link to these two Github repos/registry items now:
- https://github.com/viam-modules/detection-tflite, https://app.viam.com/ml-training/viam/detection-tflite
- https://github.com/viam-modules/classification-tflite, https://app.viam.com/ml-training/viam/classification-tflite

File: docs/services/ml/upload-training-script.md [hunk 5]
```diff
@@ -10,48 +10,231 @@ no_service: true
...
-
-{{%expand "Step 1: Parse command line inputs" %}}
+1. Create a folder for the training-script, for example <file>my-training</file>.
+1. Inside the top level folder (in this example <file>my-training</file>), create a file called `setup.py` with the following contents:
+
+   ```python {class="line-numbers linkable-line-numbers" data-line="11"}
+   from setuptools import find_packages, setup
+
+   setup(
+       name="my-training",
```
Comment (tahiyasalam) [inline docs/services/ml/upload-training-script.md:30]: super nit, but this should probably be called model

**Review by tahiyasalam — COMMENTED** [review]
Hi Naomi, this looks great! Mostly some suggestions on my end. All of the other files are not necessary and basically get autogenerated when you make the source distribution, but are not necessary for the user to create, whereas it is required the user create the setup & __init__ files

File: docs/services/ml/upload-training-script.md (hunk 1 above)
Comment (sguequierre) [reply docs/services/ml/upload-training-script.md:48]: We could also just say where to access this code snippet in app

File: docs/services/ml/upload-training-script.md (hunk 3 above)
Comment (sguequierre) [reply docs/services/ml/upload-training-script.md:167]: What would a working example look like?

File: docs/services/ml/upload-training-script.md [hunk 6]
```diff
@@ -124,154 +307,45 @@ Dataset JSON files for image datasets with bounding box labels and classificatio
...
 You must log in to the [Viam app](https://app.viam.com/) to download the package containing the classification training script from the registry.
 {{% /alert %}}
 
 {{% /expand%}}
-{{%expand "Step 4: Save the model artifact the script produces" %}}
+{{%expand "Saving the model artifact the script produces" %}}
 
-You also need to include logic in your <file>training.py</file> to save the model artifact your training job produces to the `model_output_directory`.
-For example:
+The `save_model()` and the `save_labels()` functions in the template before the `main` function saves the model artifact your training job produces to the `model_output_directory` in the cloud.
```
Comment (sguequierre) [inline docs/services/ml/upload-training-script.md:327]:
```suggestion
The `save_model()` and the `save_labels()` functions in the template before the `main` function save the model artifact your training job produces to the `model_output_directory` in the cloud.
```

File: docs/services/ml/upload-training-script.md [hunk 7]
```diff
@@ -124,154 +307,45 @@ Dataset JSON files for image datasets with bounding box labels and classificatio
...
+
+Once a training job is complete, Viam checks the output directory and creates a package with all of the contents of the directory, creating or updating a registry item for the ML model.
```
Comment (sguequierre) [inline docs/services/ml/upload-training-script.md:329]: nice

File: docs/services/ml/upload-training-script.md [hunk 8]
```diff
@@ -280,97 +354,45 @@ def save_tflite_classification(
...
+{{%expand "The main function" %}}
 
-Now, write all the code that runs the training job invoking the previously defined helper functions.
-Write this into the top level code of <file>training.py</file>, which is executed when the file runs as a script.
+The main function runs the training job by:
 
-For the [example classification training script](https://app.viam.com/packages/e76d1b3b-0468-4efd-bb7f-fb1d2b352fcb/custom-training-classification/ml_training/latest/e76d1b3b-0468-4efd-bb7f-fb1d2b352fcb) that trains a classification model using TensorFlow and Keras, `__main__` looks like this:
+1. parsing command line inputs,
+2. parsing the dataset and annotations from the dataset file,
+3. producing the model artifact, and
```
Comment (sguequierre) [inline docs/services/ml/upload-training-script.md:386]:
```suggestion
3. producing the model artifact
```
(opt) probably don't need that trailing and

File: docs/services/ml/upload-training-script.md [hunk 9]
```diff
@@ -390,7 +412,7 @@ You can use the [`viam training-script upload`](/cli/#training-script) command t
 For example:
 
 ```sh {class="command-line" data-prompt="$"}
-viam training-script upload --path=<path-to-tar.gz> --org-id=<INSERT ORG ID> --script-name="MyCustomTrainingScript"
+viam training-script upload --path=<path-to-tar.gz> --org-id=<INSERT ORG ID> --script-name=<org-id-or-namespace:training-script-name>
```
Comment (sguequierre) [inline docs/services/ml/upload-training-script.md:415]: I thought having string was helpful in example to show that's how you do it-- maybe tabs here could be helpful?

File: docs/services/ml/upload-training-script.md [hunk 10]
```diff
@@ -412,7 +434,7 @@ For example:
 {{% tab name="from-registry" %}}
 
 ```sh {class="command-line" data-prompt="$"}
-viam train submit custom from-registry --dataset-id=<INSERT DATASET ID> --org-id=<INSERT ORG ID> --model-name="MyRegistryModel" --model-version="2" --version="1" --script-name="MyCustomTrainingScript"
+viam train submit custom from-registry --dataset-id=<INSERT DATASET ID> --org-id=<INSERT ORG ID> --model-name="MyRegistryModel" --model-version="2" --version="1" --script-name="mycompany:MyCustomTrainingScript"
```
Comment (sguequierre) [inline docs/services/ml/upload-training-script.md:437]: good catch-- nvm maybe don't need tabs

File: docs/services/ml/upload-training-script.md [hunk 11]
```diff
@@ -421,7 +443,7 @@ This command submits a training job to the previously uploaded `"MyCustomTrainin
 {{% tab name="with-upload" %}}
 
 ```sh {class="command-line" data-prompt="$"}
-viam train submit custom with-upload --dataset-id=<INSERT DATASET ID> --model-org-id=<INSERT ORG ID> --model-name="MyRegistryModel" --model-type="single_label_classification" --model-version="2" --version="1" --path=<path-to-tar.gz> --script-name="MyCustomTrainingScript"
+viam train submit custom with-upload --dataset-id=<INSERT DATASET ID> --model-org-id=<INSERT ORG ID> --model-name="MyRegistryModel" --model-type="single_label_classification" --model-version="2" --version="1" --path=<path-to-tar.gz> --script-name="mycompany:MyCustomTrainingScript"
 ```
 
 This command uploads a script called `"MyCustomTrainingScript"` to the registry under the specified organization and also submits a training job to that script with the input dataset, which generates a new version of the single-classification ML model `"MyRegistryModel"` and publishes that to the registry.
```
Comment (sguequierre) [inline docs/services/ml/upload-training-script.md:449]: is the specified organization "mycompany"? ik it's org id but wondering where the mycompany comes in

**Review by sguequierre — COMMENTED** [review]
couple questions, mainly huge improvement! Thanks Naomi!

File: docs/services/ml/upload-training-script.md (hunk 3 above)
Comment (tahiyasalam) [reply docs/services/ml/upload-training-script.md:167]: [This](https://github.com/viamrobotics/docs/pull/3232/files#diff-3594b00458ec93c1fa54469b947146d47f304bedf08dbc11ec2f677e58ca13d1L212-L261) is what a working example would look like.

File: docs/services/ml/upload-training-script.md (hunk 2 above)
Comment (npentrel) [reply docs/services/ml/upload-training-script.md:125]: could you maybe suggest changes? I don't really understand what you want me to do. 
It's this button
Screenshot 2024-08-12 at 18 19 44

File: docs/services/ml/upload-training-script.md (hunk 3 above)
Comment (npentrel) [reply docs/services/ml/upload-training-script.md:167]: Yes because it's a template. I could put an example with the explanation but this code relies on other methods so it's not that simple. And it would likely get outdated and not get updated

File: docs/services/ml/upload-training-script.md (hunk 11 above)
Comment (npentrel) [reply docs/services/ml/upload-training-script.md:449]: it's the namespace or org-id I believe

File: docs/services/ml/upload-training-script.md (hunk 4 above)
Comment (npentrel) [inline docs/services/ml/upload-training-script.md:318]:
```suggestion
As an example, you can refer to the logic from the <file>model/training.py</file> from one of these training scripts:

- [detection-tflite](https://github.com/viam-modules/detection-tflite)
- [classification-tflite](https://github.com/viam-modules/classification-tflite)
```

File: docs/services/ml/upload-training-script.md [hunk 12]
```diff
@@ -124,154 +307,48 @@ Dataset JSON files for image datasets with bounding box labels and classificatio
...
-{{%expand "Step 4: Save the model artifact the script produces" %}}
+{{%expand "Saving the model artifact the script produces" %}}
 
-You also need to include logic in your <file>training.py</file> to save the model artifact your training job produces to the `model_output_directory`.
-For example:
+The `save_model()` and the `save_labels()` functions in the template before the `main` function save the model artifact your training job produces to the `model_output_directory` in the cloud.
+
+Once a training job is complete, Viam checks the output directory and creates a package with all of the contents of the directory, creating or updating a registry item for the ML model.
+
+As an example, you can refer to the logic from <file>model/training.py</file> from this [example classification training script](https://app.viam.com/packages/e76d1b3b-0468-4efd-bb7f-fb1d2b352fcb/custom-training-classification/ml_training/latest/e76d1b3b-0468-4efd-bb7f-fb1d2b352fcb) that trains a classification model using TensorFlow and Keras.
```
Comment (npentrel) [inline docs/services/ml/upload-training-script.md:334]:
```suggestion
As an example, you can refer to the logic from <file>model/training.py</file> from this [example classification training script](https://github.com/viam-modules/classification-tflite) that trains a classification model using TensorFlow and Keras.
```

File: docs/services/ml/upload-training-script.md (hunk 10 above)
Comment (npentrel) [reply docs/services/ml/upload-training-script.md:440]: ok will leave as is then for now

File: docs/services/ml/upload-training-script.md (hunk 3 above)
Comment (npentrel) [reply docs/services/ml/upload-training-script.md:167]: I've added the examples

File: docs/services/ml/upload-training-script.md [hunk 13]
```diff
@@ -10,48 +10,251 @@ no_service: true
...
+   # bounding box annotations
+   # Used for training object detection models
+   def parse_filenames_and_bboxes_from_json(
+       filename: str,
+       all_labels: ty.List[str],
+   ) -> ty.Tuple[ty.List[str], ty.List[str], ty.List[ty.List[float]]]:
+       """Load and parse JSON file to return image filenames
+       and corresponding labels with bboxes.
+       Args:
+           filename: JSONLines file containing filenames and bboxes
```
Comment (tahiyasalam) [inline docs/services/ml/upload-training-script.md:141]:
```suggestion
           filename: JSONLines file containing filenames and bboxes
           all_labels: list of all N_LABELS 
```

File: docs/services/ml/upload-training-script.md (hunk 5 above)
Comment (tahiyasalam) [reply docs/services/ml/upload-training-script.md:30]: This is the convention we use for all the example training scripts but it's up to you

File: docs/services/ml/upload-training-script.md (hunk 13 above)
Comment (npentrel) [reply docs/services/ml/upload-training-script.md:141]: ahh gotcha, thanks, I thought you meant in the actual code

File: docs/services/ml/upload-training-script.md (hunk 5 above)
Comment (npentrel) [reply docs/services/ml/upload-training-script.md:30]: It does describe the training script though right?but also it doesn't affect anything anyway, right?

**Comment by npentrel** [comment]
Ok going to merge this @tahiyasalam , if you want let me know about the name thing we were discussing I can make the change as needed in a separate PR
